### PR TITLE
Work around URL encoding in raw Lottie data

### PR DESCRIPTION
**Pull request by kaidokert**
This forces base64-encoding of all Lottie data passed in via `data:` URLs. This is needed as previous versions of Cobalt rely on non-standard `;raw,` annotation in data URLs, which isn't accepted by GURL and net::DataURL anymore.

There's an associated overhead of always forcing encoding and decoding for these, and also Parse() the URL multiple times.

However, Lottie content is relatively small and this is overhead should not be too high.

b/343971057

### CODE REVIEW

**Comment by kaidokert** [comment]
This isn't yet correct. It would re-encode already `base64` urls.

File: cobalt/dom/lottie_player.cc [hunk 1]
```diff
@@ -298,9 +299,28 @@ void LottiePlayer::UpdateAnimationData() {
...
+        return beforeComma.find("base64") != std::string::npos;
+      }
+      return false;
+    };
+
+    // Force base64 encoding of plain/raw inputs, to ensure no URL escaping
+    // occurs.
+    GURL resource_url = selected_source;
+    if (selected_source.SchemeIs("data") && !isBase64Encoded(src)) {
+      auto encoded = ForgivingBase64Encode(src.substr(src.find(',') + 1));
```
Comment (andrewsavage1) [inline cobalt/dom/lottie_player.cc:315]: nit: I personally find it hard to read code that uses `auto` then dereferences it

File: cobalt/dom/lottie_player.cc [hunk 2]
```diff
@@ -298,9 +299,28 @@ void LottiePlayer::UpdateAnimationData() {
...
+    auto isBase64Encoded = [](const std::string& url) -> bool {
+      size_t pos = url.find(',');
+      if (pos != std::string::npos) {
+        std::string beforeComma = url.substr(0, pos);
+        return beforeComma.find("base64") != std::string::npos;
+      }
+      return false;
+    };
+
+    // Force base64 encoding of plain/raw inputs, to ensure no URL escaping
```
Comment (andrewsavage1) [inline cobalt/dom/lottie_player.cc:311]: Should there be a TODO with a bug to remove the non-standard urls we use in the future?

**Review by andrewsavage1 — APPROVED** [review]

File: cobalt/dom/lottie_player.cc [hunk 3]
```diff
@@ -298,9 +299,28 @@ void LottiePlayer::UpdateAnimationData() {
       return;
     }
 
+    auto isBase64Encoded = [](const std::string& url) -> bool {
+      size_t pos = url.find(',');
+      if (pos != std::string::npos) {
+        std::string beforeComma = url.substr(0, pos);
+        return beforeComma.find("base64") != std::string::npos;
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply cobalt/dom/lottie_player.cc:306]: I'm not sure how whitespace is treated, and how client code passes it. It's possible some of the callers pass in spaces here and so i'd be as lenient as possible here. 

I initially used `net/base/data_url.h` here to delegate full parsing to, but that breaks further as it won't deal with non-standard `;raw,` input at all.

I'd leave this as conservative as possible here.

**Review by jellefoks — APPROVED** [review]

**Review by briantting — APPROVED** [review]

File: cobalt/dom/lottie_player.cc [hunk 4]
```diff
@@ -298,9 +299,28 @@ void LottiePlayer::UpdateAnimationData() {
...
+      }
+      return false;
+    };
+
+    // Force base64 encoding of plain/raw inputs, to ensure no URL escaping
+    // occurs.
+    GURL resource_url = selected_source;
```
[the remark this answers is not in the document]
Comment (briantting) [reply cobalt/dom/lottie_player.cc:313]: +1 on having this be its own function alongside IsBase64Encoded

File: cobalt/dom/lottie_player.cc [hunk 5]
```diff
@@ -298,9 +299,28 @@ void LottiePlayer::UpdateAnimationData() {
       return;
     }
 
+    auto isBase64Encoded = [](const std::string& url) -> bool {
```
[the remark this answers is not in the document]
Comment (kaidokert) [reply cobalt/dom/lottie_player.cc:302]: Moved

File: cobalt/dom/lottie_player.cc (hunk 1 above)
Comment (kaidokert) [reply cobalt/dom/lottie_player.cc:315]: Changed the return type to optional, it's a lot more verbose.

Comment (andrewsavage1) [reply cobalt/dom/lottie_player.cc]: If this is an optional and you don't like the verbosity you can communicate what it is by checking `encoded.has_value()` and `encoded.value()` in place of of `encoded`/`*encoded`

File: cobalt/dom/lottie_player.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (kaidokert) [reply cobalt/dom/lottie_player.cc:315]: Leaving as is for now then.

[the remark this answers is not in the document]
Comment (andrewsavage1) [reply cobalt/dom/lottie_player.cc:315]: I'll just say I disagree with this as a function would be better for self-documentation (as opposed to the comment which will eventually be outdated) and makes the code easier to understand with a good name

**Pull request merged**